ranger-discovery-bundle/src/main/java/io/appform/ranger/discovery/bundle/id/nonce/RandomNonceGenerator.java

```diff
@@ -0,0 +1,139 @@
+        val collisionChecker = !Strings.isNullOrEmpty(request.getDomain())
+                ? getREGISTERED_DOMAINS().getOrDefault(request.getDomain(), Domain.DEFAULT)
+                .getCollisionChecker()
+                : Domain.DEFAULT.getCollisionChecker();
+        return Optional.ofNullable(RETRYER.get(
+                () -> {
+                    IdInfo idInfo = random(collisionChecker);
+                    val id = getIdFromIdInfo(idInfo, request.getPrefix(), request.getIdFormatter());
+                    return GenerationResult.builder()
+                            .idInfo(idInfo)
+                            .state(validateId(request.getConstraints(), id, request.isSkipGlobal()))
+                            .domain(null)
+                            .build();
+                }))
+                .filter(generationResult -> generationResult.getState() == IdValidationState.VALID)
+                .map(GenerationResult::getIdInfo);
+    }
+
+    @Override
+    public Optional<IdInfo> generateWithConstraints(final String namespace, final List<IdValidationConstraint> inConstraints, final boolean skipGlobal) {
+        return Optional.ofNullable(RETRYER.get(
+                        () -> {
+                            val idInfo = generate(namespace);
+                            val id = getIdFromIdInfo(idInfo, namespace, getIdFormatter());
+                            return GenerationResult.builder()
+                                    .idInfo(idInfo)
+                                    .state(validateId(inConstraints, id, skipGlobal))
+                                    .domain(null)
+                                    .build();
+                        }))
+                .filter(generationResult -> generationResult.getState() == IdValidationState.VALID)
+                .map(GenerationResult::getIdInfo);
+    }
+
+    private IdInfo random(final CollisionChecker collisionChecker) {
+        int randomGen;
+        long time;
+        do {
+            time = System.currentTimeMillis();
+            randomGen = getSECURE_RANDOM().nextInt(Constants.MAX_ID_PER_MS);
+        } while (!collisionChecker.check(time, randomGen));
+        return new IdInfo(randomGen, time);
+    }
+
+    private static int readRetryCount() {
+        try {
+            val count = Integer.parseInt(System.getenv().getOrDefault("NUM_ID_GENERATION_RETRIES", "512"));
+            if (count <= 0) {
+                throw new IllegalArgumentException(
+                        "Negative number of retries does not make sense. Please set a proper value for " +
+                                "NUM_ID_GENERATION_RETRIES");
+            }
+            return count;
+        }
+        catch (NumberFormatException e) {
+            throw new IllegalArgumentException("Please provide a valid positive integer for NUM_ID_GENERATION_RETRIES");
+        }
+    }
+
+    @Override
+    public DateTime getDateTimeFromTime(final long time) {
```

Comment:
what is this method, why is this public and why is it placed below private methods?